airbyte-integrations/connectors/destination-s3/src/main/java/io/airbyte/integrations/destination/s3/csv/S3CsvOutputFormatter.java

```diff
@@ -0,0 +1,269 @@
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.nio.charset.StandardCharsets;
+import java.sql.Timestamp;
+import java.util.Collections;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.UUID;
+import java.util.stream.Collectors;
+import org.apache.commons.csv.CSVFormat;
+import org.apache.commons.csv.CSVPrinter;
+import org.apache.commons.csv.QuoteMode;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class S3CsvOutputFormatter implements S3OutputFormatter {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(S3CsvOutputFormatter.class);
+  private static final ExtendedNameTransformer NAME_TRANSFORMER = new ExtendedNameTransformer();
+
+  private final S3DestinationConfig config;
+  private final S3CsvFormatConfig formatConfig;
+  private final AmazonS3 s3Client;
+  private final AirbyteStream stream;
+  private final DestinationSyncMode syncMode;
+  private final List<String> sortedHeaders;
+  private final String outputPrefix;
+  private final StreamTransferManager uploadManager;
+  private final MultiPartOutputStream outputStream;
+  private final CSVPrinter csvPrinter;
+
+  public S3CsvOutputFormatter(S3DestinationConfig config,
+                              AmazonS3 s3Client,
+                              ConfiguredAirbyteStream configuredStream,
+                              Timestamp uploadTimestamp)
+      throws IOException {
+    this.config = config;
+    this.formatConfig = (S3CsvFormatConfig) config.getFormatConfig();
+    this.s3Client = s3Client;
+    this.stream = configuredStream.getStream();
+    this.syncMode = configuredStream.getDestinationSyncMode();
+    this.sortedHeaders = getSortedFields(configuredStream.getStream().getJsonSchema(),
+        formatConfig);
+
+    // prefix: <bucket-path>/<source-namespace-if-exists>/<stream-name>
+    // filename: <upload-date>-<upload-millis>.csv
+    // full path: <bucket-name>/<prefix>/<filename>
+    this.outputPrefix = getOutputPrefix(config.getBucketPath(), stream);
+    String outputFilename = getOutputFilename(uploadTimestamp);
+    String objectKey = String.join("/", outputPrefix, outputFilename);
+
+    LOGGER.info("Full S3 path for stream '{}': {}/{}", stream.getName(), config.getBucketName(),
+        objectKey);
+
+    // The stream transfer manager lets us greedily stream into S3. The native AWS SDK does not
+    // have support for streaming multipart uploads. The alternative is first writing the entire
+    // output to disk before loading into S3. This is not feasible with large input.
+    // Data is chunked into parts during the upload. A part is sent off to a queue to be uploaded
+    // once it has reached it's configured part size.
+    // Memory consumption is queue capacity * part size = 1 * 5 = 5 MB at current configurations.
+    this.uploadManager = new StreamTransferManager(config.getBucketName(), objectKey,
+        s3Client)
+            .numUploadThreads(1)
```

Comment:
Would adding more threads help performance? What if we aim for somewhere between 300-500mb of usage? Nothing scientific backing this - just finger to the wind estimate. Right now it feels we are being too conservative.

Reply:
Alright, I increased number of threads and queue size to 2. The current memory usage per stream is 31 MB. I am conservative about these parameters because it is possible to have many streams per data sync, which can result in OOME.